src/rust/runtime/scheduler/scheduler.rs

```diff
@@ -61,197 +42,78 @@ pub struct SharedScheduler(SharedObject<Scheduler>);
+            if let Some(task) = group.poll_notified_task_and_remove_if_ready(id) {
                 completed_tasks.push(task);
             }
         }
         completed_tasks
     }
 
-    /// Poll all tasks until one completes. Remove that task and return it or fail after polling [max_iteration] number
-    /// of tasks.
-    pub fn get_next_completed_task(&mut self, max_iterations: usize) -> Option<Box<dyn Task>> {
-        for _ in 0..max_iterations {
-            self.current_task_id = {
-                match self.current_ready_tasks.pop() {
-                    Some(index) => index,
-                    None => {
-                        self.next_runnable_group();
-                        match self.current_ready_tasks.pop() {
-                            Some(index) => index,
-                            None => return None,
-                        }
-                    },
-                }
-            };
-
-            // Now that we have a runnable task, actually poll it.
-            if let Some(task) = self.poll_notified_task_and_remove_if_ready() {
-                return Some(task);
-            }
-        }
-        None
-    }
-
-    /// Poll over all of the groups looking for a group with runnable tasks. Sets the current_group_id to the next
-    /// runnable task group and current_ready_tasks to a list of tasks that are runnable in that group.
-    fn next_runnable_group(&mut self) {
-        let starting_group_index: InternalId = self.current_group_id;
-        self.current_group_id = self.get_next_group_index();
-
-        loop {
-            self.current_ready_tasks = self.groups[self.current_group_id.into()].get_offsets_for_ready_tasks();
-            if !self.current_ready_tasks.is_empty() {
-                return;
+    pub fn poll_group_until_unrunnable(&mut self, group_id: TaskId, max_iterations: usize) -> Vec<Box<dyn Task>> {
+        let mut completed_tasks: Vec<Box<dyn Task>> = vec![];
+        // Keep running as long as there are runnable tasks.
+        let mut iterations: usize = 0;
+        // Expect is safe here because something has really gone wrong if we are polling a group that doesn't exist.
+        let group: &mut TaskGroup = self.get_mut_group(group_id).expect("group being polled doesn't exist");
+        while iterations < max_iterations {
+            let ready_tasks: Vec<InternalId> = group.get_offsets_for_ready_tasks();
+            if ready_tasks.is_empty() {
+                break;
             }
-            // If we reach this point, then we have looped all the way around without finding any runnable tasks.
-            if self.current_group_id == starting_group_index {
-                return;
+            for id in ready_tasks {
+                // Now that we have a runnable task, actually poll it.
+                if let Some(task) = group.poll_notified_task_and_remove_if_ready(id) {
+                    completed_tasks.push(task);
+                }
+                iterations += 1;
```

Comment:
Do we want to run until there are `max_iteration` number of ready tasks, or `max_iteration` number of times? Here, it seems to count the number of tasks that are ready.